data/provider/src/main/java/tech/pegasys/teku/api/ChainDataProvider.java

```diff
@@ -684,14 +687,80 @@ protected SyncCommitteeRewardData calculateRewards(
     return data;
   }
 
+  public SafeFuture<Optional<ObjectAndMetaData<BlockRewardData>>> getBlockRewardsFromBlockId(
+      final String blockId) {
+    return getBlockAndMetaData(blockId)
+        .thenCompose(
+            result -> {
+              if (result.isEmpty() || result.get().getData().getBeaconBlock().isEmpty()) {
+                return SafeFuture.completedFuture(Optional.empty());
+              }
+              final BlockAndMetaData blockAndMetaData = result.get();
+              final BeaconBlock block = blockAndMetaData.getData().getBeaconBlock().get();
+
+              return combinedChainDataClient
+                  .getStateByBlockRoot(block.getRoot())
+                  .thenApply(
+                      maybeState ->
+                          maybeState.map(
+                              state -> getBlockRewardData(blockAndMetaData, block, state)));
+            });
+  }
+
+  private ObjectAndMetaData<BlockRewardData> getBlockRewardData(
+      final BlockAndMetaData blockAndMetaData,
+      final BeaconBlock block,
+      final tech.pegasys.teku.spec.datastructures.state.beaconstate.BeaconState state) {
+    if (!spec.atSlot(block.getSlot()).getMilestone().isGreaterThanOrEqualTo(SpecMilestone.ALTAIR)) {
+      throw new BadRequestException(
+          "Slot "
+              + block.getSlot()
+              + " is pre altair, and no sync committee information is available");
+    }
+
+    final UInt64 participantReward = spec.getSyncCommitteeParticipantReward(state);
+    final UInt64 proposerReward =
+        participantReward
+            .times(PROPOSER_WEIGHT)
+            .dividedBy(WEIGHT_DENOMINATOR.minus(PROPOSER_WEIGHT));
+    return blockAndMetaData.map(value -> calculateBlockRewards(proposerReward, block, state));
+  }
+
+  private BlockRewardData calculateBlockRewards(
+      final UInt64 proposerReward,
+      final BeaconBlock block,
+      final tech.pegasys.teku.spec.datastructures.state.beaconstate.BeaconState state) {
+    final SyncAggregate aggregate = block.getBody().getOptionalSyncAggregate().orElseThrow();
+
+    final int proposerIndex = spec.getBeaconProposerIndex(state, block.getSlot());
+    final long attestationsBlockRewards = calculateAttestationRewards(state);
+    final UInt64 syncAggregateBlockRewards =
+        calculateProposerSyncAggregateBlockRewards(proposerReward, aggregate);
+    final UInt64 proposerSlashingsBlockRewards = calculateProposerSlashingsRewards(block, state);
+    final UInt64 attesterSlashingsBlockRewards = calculateAttesterSlashingsRewards(block, state);
+    final long total =
+        attestationsBlockRewards
+            + syncAggregateBlockRewards.longValue()
+            + proposerSlashingsBlockRewards.longValue()
+            + attesterSlashingsBlockRewards.longValue();
```

Comment:
I guess I did it that way, because thats just adding up the total, and the other values are supposed to be UInt64? I could change that if we wanted to change the structure of the output data.

Reply:
there's .longValue everywhere which suggests maybe worth taking a look... if we start with primitives it may be simpler in some ways